opencsp/app/fixed_pattern_deflectometry/lib/FixedPatternSetupCalibrate.py

```diff
@@ -178,21 +189,51 @@ def _find_markers_in_images(self) -> None:
                 self.marker_detection_params['adaptive_thresh_constant'],
                 self.marker_detection_params['min_marker_perimeter_rate'],
             )
-            # Save point locations and IDs
+            # Save corner locations and corner IDs
             marker_ids = np.repeat(ids, 4)
             marker_corner_ids = np.repeat(ids * 4, 4) + np.tile(ids_add, ids.size)
             marker_corners_xy = Vxy(np.concatenate(pts, 0).T)
-            # Save xyz locations
-            point_idxs = []
-            for marker_corner_id in marker_corner_ids:
-                point_idxs.append(
-                    np.where(self._pts_ids_corners == marker_corner_id)[0][0]
-                )
 
-            self._marker_ids.append(marker_ids)
-            self._marker_corner_ids.append(marker_corner_ids)
-            self._marker_corners_xy.append(marker_corners_xy)
-            self._marker_corners_xyz.append(self._pts_xyz_corners[point_idxs])
+            # Save xyz locations
+            marker_corner_ids_cur = []
+            marker_ids_cur = []
+            marker_corner_xy_cur = []
+            marker_corner_xyz_cur = []
+            marker_corner_idxs_cur = []
+            for marker_corner_id, marker_id, marker_corner_xy in zip(marker_corner_ids, marker_ids, marker_corners_xy):
+                # Loop up found marker corner ID
+                idx_mkr_corner = np.where(
+                    self._pts_ids_corners == marker_corner_id)[0]
+
+                # Check marker corner ID is in given list of marker corner IDs
+                if idx_mkr_corner.size == 0:
+                    if self._to_print:
+                        print(
+                            f'Found marker corner ID, {marker_corner_id:d}, is not in given marker corner IDs. This point will be skipped.')
+                    continue
+
+                # Check found marker corner ID is assigned a non-nan value in given marker locations
+                if np.any(np.isnan(self._pts_xyz_corners[idx_mkr_corner[0]].data)):
+                    if self._to_print:
+                        print(
+                            f'Point ID {idx_mkr_corner[0]:d} undefined, point will be skipped')
+                    continue
+
+                # Save marker corner ID
+                marker_corner_idxs_cur.append(idx_mkr_corner[0])
+                marker_corner_ids_cur.append(marker_corner_id)
+                marker_ids_cur.append(marker_id)
+                marker_corner_xy_cur.append(marker_corner_xy.data)
+                marker_corner_xyz_cur.append(
+                    self._pts_xyz_corners[idx_mkr_corner[0]].data)
+
+            self._marker_corner_idxs.append(np.array(marker_corner_idxs_cur))
```

Comment:
All this related book keeping information (indexes, ids, xy locations, etc...) seems like an easy place to make a mistake and introduce bugs.  Can it be better encapsulated in a "Marker" class?  Or with a "add_marker()" method?

Reply:
There is technically a Marker (common/lib/photogrammetry/ImageMarker.py) class that I didn't think about using here. I think that would be a good idea.

Reply:
Aha, that would explain why I didn't see it as a new class in your change :)